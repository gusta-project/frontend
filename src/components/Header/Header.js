import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import React, { Component, Fragment } from 'react';
import { bindActionCreators } from 'redux';
import { NavLink } from 'react-router-dom';
import { Navbar, Nav, NavDropdown, Container, Row, Col } from 'react-bootstrap';

import { actions as appActions } from 'reducers/application';
import { isLoggedIn } from 'selectors/application';

export class Header extends Component {
  static propTypes = {
    actions: PropTypes.object.isRequired,
    loggedIn: PropTypes.bool.isRequired
  };
  constructor(props) {
    super(props);

    this.logoutUser = this.logoutUser.bind(this);
  }
  renderNavItem(to, text) {
    return (
      <Nav.Link
        as={NavLink}
        exact
        to={to}
        className="nav--link-custom px-3"
        activeClassName="active"
      >
        {text}
      </Nav.Link>
    );
  }

  renderNavDropdownItem(to, text) {
    return (
      <NavDropdown.Item
        as={NavLink}
        exact
        to={to}
        className="nav--link-custom"
        activeClassName="active"
      >
        {text}
        <br />
      </NavDropdown.Item>
    );
  }

  logoutUser() {
    const { actions } = this.props;

    actions.logoutUser();
  }

  render() {
    const { loggedIn } = this.props;

    return (
      <Container fluid>
        <Row className="navigation-container">
          <Col>
            <Navbar expand="lg">
              <Navbar.Brand className="pt-0">
                <img src="/media/logo.svg" alt="logo" className="image--logo" />
              </Navbar.Brand>
              <Navbar.Toggle aria-controls="basic-navbar-nav" />
              <Navbar.Collapse>
                <Nav>
                  {this.renderNavItem('/', 'Home')}
                  {this.renderNavItem('/recipes', 'Recipes')}
                  {loggedIn
                    ? this.renderNavItem('/calculator', 'Calculator')
                    : null}
                  {this.renderNavItem('/flavors', 'Flavors')}
                  {loggedIn ? (
                    <NavDropdown title="User">
                      {this.renderNavDropdownItem('/user/profile', 'Profile')}
                      {this.renderNavDropdownItem('/user/recipes', 'Recipes')}
                      {this.renderNavDropdownItem(
                        '/user/favorites',
                        'Favorites'
                      )}
                      {this.renderNavDropdownItem(
                        '/user/flavor-stash',
                        'Flavor Stash'
                      )}
                      {this.renderNavDropdownItem(
                        '/user/shopping-list',
                        'Shopping List'
                      )}
                      {this.renderNavDropdownItem('/user/settings', 'Settings')}

                      {this.renderNavDropdownItem('/dashboard', 'Dashboard')}
                      <NavDropdown.Item
                        as={NavLink}
                        to="#"
                        className="nav--link-custom"
                        activeClassName="active"
                        onClick={this.logoutUser}
                      >
                        Logout
                        <br />
                      </NavDropdown.Item>
                    </NavDropdown>
                  ) : null}
                  {!loggedIn ? (
                    <Fragment>
                      {this.renderNavItem('/login', 'Login')}
                      {this.renderNavItem('/register', 'Register')}
                    </Fragment>
                  ) : null}
                </Nav>
              </Navbar.Collapse>
            </Navbar>
          </Col>
        </Row>
      </Container>
    );
  }
}

const mapStateToProps = state => ({
  loggedIn: isLoggedIn(state)
});

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators(appActions, dispatch)
});

export default connect(mapStateToProps, mapDispatchToProps)(Header);
