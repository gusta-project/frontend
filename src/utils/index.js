import dayjs from 'dayjs';

import { getApplication, getAuthorization } from 'selectors/application';

/* eslint-disable react/display-name, react/prop-types */

/**
 * Creates an object containing action constants namespaced under the specified reducer.
 *
 * @param {string} reducer The name of the reducer or module
 * @param {string[]} actions A list of action names
 * @return {object} Object with action name keys and full action string values
 */
export function buildActions(reducer, actions) {
  const result = {};

  for (const action of actions) {
    result[action] = `${reducer}/${action}`;
  }

  return result;
}

/**
 * Interpolate strings from the `params` object into the `url`.
 *
 * @param {object} endpoint Object with `url` and optional `params` object
 */
export const buildUrl = (endpoint) => {
  // this is defined by webpack.DefinePlugin
  const baseUrl = API_URL;
  const { url, params } = endpoint;

  if (!params) {
    return `${baseUrl}${url}`;
  }

  return Object.entries(params).reduce((resultUrl, entry) => {
    const [key, value] = entry;

    return resultUrl.replace(`{${key}}`, value);
  }, `${baseUrl}${url}`);
};

/**
 * Gets the initial Redux state from localStorage and the source code defaults.
 */
export const getInitialState = () => {
  // this needs to be require()d because an import results in a
  // circular dependency
  const { initialState } = require('reducers');

  try {
    const [accessToken, expiration] = [
      localStorage.getItem('accessToken'),
      localStorage.getItem('expiration')
    ];

    if (!accessToken) {
      return initialState;
    }

    const expirationDate = dayjs(JSON.parse(expiration));

    if (!expirationDate.isValid() || expirationDate.isBefore(dayjs())) {
      localStorage.removeItem('accessToken');
      localStorage.removeItem('expiration');
      return initialState;
    }

    return {
      ...initialState,
      application: {
        ...getApplication(initialState),
        authorization: {
          ...getAuthorization(initialState),
          accessToken: JSON.parse(accessToken),
          expiration: expirationDate
        }
      }
    };
  } catch {
    return initialState;
  }
};
