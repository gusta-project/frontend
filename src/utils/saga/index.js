import counterUtilities from './counter';
import pageUtilities from './pager';
import toastUtilities from './toasts';

export default {
  ...counterUtilities,
  ...pageUtilities,
  ...toastUtilities
};
