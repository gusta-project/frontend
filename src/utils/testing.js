import React, { createElement } from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';

/**
 * Creates a mock component which takes a set of props specified at runtime.
 *
 * @param {string} name The component/element name (in PascalCase)
 * @param {object} props An object containing props to setup
 * @return {object} Mock component with specified name and props
 */
export const mockComponent = (name, props = {}) => () =>
  createElement(name, props, props.children);

/**
 * Creates a mock component which will expose its props for snapshot testing purposes
 * @param {*} name The component/element name (in PascalCase)
 */
export const mockComponentWithProps = (name) => (props) =>
  createElement(name, props, props.children);

/**
 * Wraps a React component in a <MemoryRouter> suitable for testing.
 *
 * @param {Component} WrappedComponent React component to wrap
 */
export const withMemoryRouter = (WrappedComponent, routerProps = {}) => (
  props
) => (
  <MemoryRouter {...routerProps}>
    <WrappedComponent {...props} />
  </MemoryRouter>
);

/**
 * Wraps a React component in a <Provider> suitable for testing.
 *
 * @param {Component} WrappedComponent React component to wrap
 * @param {object} store Mock store
 */
export const withProvider = (WrappedComponent, store) => (props) => (
  <Provider store={store}>
    <WrappedComponent {...props} />
  </Provider>
);
