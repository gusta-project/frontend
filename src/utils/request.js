import axios from 'axios';
import { buildUrl } from 'utils';
import { call, race, delay } from 'redux-saga/effects';

/**
 * Make an XMLHttpRequest to the specified endpoint with the specified request data.
 *
 * @param {object} endpoint An object with `url` and `method` keys
 * @param {object} data An object to use as the request body
 */
class Request {
  *execute({ endpoint, data, options = {} }) {
    try {
      // ensure the endpoint was supplied
      if (!endpoint) {
        throw new Error('No endpoint provided!');
      }

      // ensure properties required by axios are present
      const { url, method } = endpoint;

      if (!url) {
        throw new Error('Endpoint is missing URL!');
      }

      if (!method) {
        throw new Error('Endpoint is missing method!');
      }

      // default timeout of 30 seconds
      const { timeout = 30000 } = options;
      const requestUrl = yield call(buildUrl, endpoint);

      // start a race between the request and a timer, cancel the loser
      const { response } = yield race({
        response: call(axios, {
          url: requestUrl,
          method,
          data
        }),
        timeout: delay(timeout)
      });

      if (!response) {
        throw new Error('Request timed out!');
      }

      const { status } = response;
      const success = status >= 200 && status < 400;

      if (success) {
        return {
          success,
          response
        };
      } else {
        return {
          success,
          error: response
        };
      }
    } catch (error) {
      return {
        success: false,
        error
      };
    }
  }
}

const instance = new Request();

export default instance;
