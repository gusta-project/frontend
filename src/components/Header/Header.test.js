import React from 'react';
import renderer from 'react-test-renderer';
import { Header } from './Header';
import { withMemoryRouter } from 'utils/testing';
// Prevent findDOMNode error in test from Dropdown component in react-bootstrap
jest.mock('react-dom', () => ({
  findDOMNode: () => ({})
}));

describe('<Header />', () => {
  const actions = {
    logoutUser: jest.fn(),
    requestCurrentUser: jest.fn()
  };
  const RoutedHeader = withMemoryRouter(Header);

  it('renders logged out correctly', () => {
    const props = {
      actions,
      loggedIn: false
    };

    expect(
      renderer.create(<RoutedHeader {...props} />).toJSON()
    ).toMatchSnapshot();
  });

  it('renders logged in correctly', () => {
    const props = {
      actions,
      loggedIn: true
    };

    expect(
      renderer.create(<RoutedHeader {...props} />).toJSON()
    ).toMatchSnapshot();
  });

  it('can logoutUser', () => {
    const component = renderer.create(
      <RoutedHeader actions={actions} loggedIn />
    );
    const { instance } = component.root.findByType(Header);

    expect(instance).toBeDefined();
    instance.logoutUser();
    expect(actions.logoutUser).toHaveBeenCalledWith();
  });

  it('can render gravatar', () => {
    const props = {
      actions,
      loggedIn: true,
      user: { id: 1, emailAddress: 'jest@mixnjuice.com' }
    };
    const component = renderer.create(<RoutedHeader {...props} />);
    const { instance } = component.root.findByType(Header);

    expect(instance).toBeDefined();
    expect(instance.gravatar).toEqual(
      'https://www.gravatar.com/avatar/6114a57951ce2eefc79a1b3d7918a6f5?s=50&d=mp'
    );
  });
});
