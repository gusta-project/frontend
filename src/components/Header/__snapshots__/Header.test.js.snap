// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> renders logged in correctly 1`] = `
<div
  className="mb-5 container-fluid"
>
  <div
    className="navigation-container row"
  >
    <div
      className="col"
    >
      <nav
        className="navbar navbar-expand-lg navbar-light"
      >
        <span
          className="pt-0 navbar-brand"
        >
          <img
            alt="logo"
            className="image--logo"
            src="/media/logo.svg"
          />
        </span>
        <button
          aria-controls="basic-navbar-nav"
          aria-label="Toggle navigation"
          className="navbar-toggler collapsed"
          onClick={[Function]}
          type="button"
        >
          <span
            className="navbar-toggler-icon"
          />
        </button>
        <div
          aria-expanded={null}
          className="navbar-collapse collapse"
        >
          <div
            className="navbar-nav"
            onKeyDown={[Function]}
          >
            <a
              aria-current="page"
              className="nav--link-custom px-3 nav-link active"
              data-rb-event-key={null}
              disabled={false}
              href="/"
              onClick={[Function]}
              style={Object {}}
            >
              Home
            </a>
            <a
              aria-current={null}
              className="nav--link-custom px-3 nav-link"
              data-rb-event-key={null}
              disabled={false}
              href="/recipes"
              onClick={[Function]}
            >
              Recipes
            </a>
            <a
              aria-current={null}
              className="nav--link-custom px-3 nav-link"
              data-rb-event-key={null}
              disabled={false}
              href="/recipe/editor"
              onClick={[Function]}
            >
              Recipe Editor
            </a>
            <a
              aria-current={null}
              className="nav--link-custom px-3 nav-link"
              data-rb-event-key={null}
              disabled={false}
              href="/flavors"
              onClick={[Function]}
            >
              Flavors
            </a>
          </div>
          <div
            className="ml-auto navbar-nav"
            onKeyDown={[Function]}
          >
            <div
              className="nav--link-custom px-3 dropdown"
              onKeyDown={[Function]}
            >
              <a
                aria-expanded={false}
                aria-haspopup={true}
                className="dropdown-toggle"
                href="/"
                id="user-dropdown"
                onClick={[Function]}
              />
            </div>
          </div>
        </div>
      </nav>
    </div>
  </div>
</div>
`;

exports[`<Header /> renders logged out correctly 1`] = `
<div
  className="mb-5 container-fluid"
>
  <div
    className="navigation-container row"
  >
    <div
      className="col"
    >
      <nav
        className="navbar navbar-expand-lg navbar-light"
      >
        <span
          className="pt-0 navbar-brand"
        >
          <img
            alt="logo"
            className="image--logo"
            src="/media/logo.svg"
          />
        </span>
        <button
          aria-controls="basic-navbar-nav"
          aria-label="Toggle navigation"
          className="navbar-toggler collapsed"
          onClick={[Function]}
          type="button"
        >
          <span
            className="navbar-toggler-icon"
          />
        </button>
        <div
          aria-expanded={null}
          className="navbar-collapse collapse"
        >
          <div
            className="navbar-nav"
            onKeyDown={[Function]}
          >
            <a
              aria-current="page"
              className="nav--link-custom px-3 nav-link active"
              data-rb-event-key={null}
              disabled={false}
              href="/"
              onClick={[Function]}
              style={Object {}}
            >
              Home
            </a>
            <a
              aria-current={null}
              className="nav--link-custom px-3 nav-link"
              data-rb-event-key={null}
              disabled={false}
              href="/recipes"
              onClick={[Function]}
            >
              Recipes
            </a>
            <a
              aria-current={null}
              className="nav--link-custom px-3 nav-link"
              data-rb-event-key={null}
              disabled={false}
              href="/flavors"
              onClick={[Function]}
            >
              Flavors
            </a>
            <a
              aria-current={null}
              className="nav--link-custom px-3 nav-link"
              data-rb-event-key={null}
              disabled={false}
              href="/login"
              onClick={[Function]}
            >
              Login
            </a>
            <a
              aria-current={null}
              className="nav--link-custom px-3 nav-link"
              data-rb-event-key={null}
              disabled={false}
              href="/register"
              onClick={[Function]}
            >
              Register
            </a>
          </div>
          <div
            className="ml-auto navbar-nav"
            onKeyDown={[Function]}
          />
        </div>
      </nav>
    </div>
  </div>
</div>
`;
